### PR TITLE
[bitnami/thanos] Add customAutoScaling flag

### PR DESCRIPTION
**Pull request by bingikarthik**
### Description of the change

Add the `customAutoScaling` flag to allow users/teams using CustomHPA for `thanos-query` to disable the replicas field in the deployment. Currently, the `thanos-query` app is getting out of sync due to a conflict between HPA and ArgoCD when trying to adjust the deployment replicas.

image

### CODE REVIEW

**Comment by carrodher** [comment]
Thanks for your contribution! Please note you need to add the metadata to automatically generate the documentation for any new parameter in the values.yaml, see the already existing metadata for similar parameters:
```yaml
  ## @param query.autoscaling.enabled Enable autoscaling for Thanos Query
  ## @param query.autoscaling.minReplicas Minimum number of Thanos Query replicas
  ## @param query.autoscaling.maxReplicas Maximum number of Thanos Query replicas
  ## @param query.autoscaling.targetCPU Target CPU utilization percentage
  ## @param query.autoscaling.targetMemory Target Memory utilization percentage
  ##
  autoscaling:
    enabled: false
    minReplicas: ""
    maxReplicas: ""
    targetCPU: ""
    targetMemory: ""
```

**Comment by bingikarthik** [comment]
> Thanks for your contribution! Please note you need to add the metadata to automatically generate the documentation for any new parameter in the values.yaml, see the already existing metadata for similar parameters:
> 
> ```yaml
>   ## @param query.autoscaling.enabled Enable autoscaling for Thanos Query
>   ## @param query.autoscaling.minReplicas Minimum number of Thanos Query replicas
>   ## @param query.autoscaling.maxReplicas Maximum number of Thanos Query replicas
>   ## @param query.autoscaling.targetCPU Target CPU utilization percentage
>   ## @param query.autoscaling.targetMemory Target Memory utilization percentage
>   ##
>   autoscaling:
>     enabled: false
>     minReplicas: ""
>     maxReplicas: ""
>     targetCPU: ""
>     targetMemory: ""
> ```

Thanks, @carrodher, for the detailed explanation. The metadata has now been added for the Thanos Query Custom Autoscaling configuration.

**Comment by bingikarthik** [comment]
> Hi @bingikarthik! Thanks so much for your contribution.
By CustomHPA do you refer to [this particular project](https://github.com/nanit/kubernetes-custom-hpa) or any other similar one?

Hi @juan131,
I’m referring to the fact that we use a custom HPA instead of the generic one defined here: https://github.com/bitnami/charts/blob/main/bitnami/thanos/templates/query/hpa.yaml, which is based on CPU and memory scaling.
Our custom hpa, would look similar to this:
```
apiVersion: autoscaling/v2
kind: HorizontalPodAutoscaler
metadata:
  name: {{ include "common.names.fullname" $thanosValues }}-query
  namespace: {{ .Release.Namespace }}
  labels:
    {{- include "ethos-core.labels" . | nindent 4 }}
spec:
  scaleTargetRef:
    apiVersion: apps/v1
    kind: Deployment
    name: {{ include "common.names.fullname" $thanosValues }}-query
  minReplicas: {{ .Values.thanos.query.autoscaling.minReplicas | default 2 }}
  maxReplicas: {{ .Values.thanos.query.autoscaling.maxReplicas | default 10 }}
  metrics:
    - type: Pods
      pods:
        metric:
          name: thanos_query_range_query_duration
        target:
          type: AverageValue
          averageValue: {{ .Values.thanos.query.autoscaling.thanosQueryRangeQueryDuration | default 2 }}
    - type: Pods
      pods:
        metric:
          name: thanos_query_instant_query_duration
        target:
          type: AverageValue
          averageValue: {{ .Values.thanos.query.autoscaling.thanosQueryInstantQueryDuration | default 0.7 }}
    - type: Pods
      pods:
        metric:
          name: thanos_query_promql_duration_metric
        target:
          type: AverageValue
          averageValue: {{ .Values.thanos.query.autoscaling.thanosQueryPromqlDurationMetric | default 0.5 }}
```

**Comment by juan131** [comment]
Thanks for the context @bingikarthik ! Then I think we'll should use a different approach and provide you a way to customize the existing HPA in the chart so it's less opinionated. What do you think?

**Comment by bingikarthik** [comment]
> Thanks for the context @bingikarthik ! Then I think we'll should use a different approach and provide you a way to customize the existing HPA in the chart so it's less opinionated. What do you think?

Agree, that's make sense!
I will make the changes accordingly to the existing HPA only.

**Comment by bingikarthik** [comment]
> Thanks for the context @bingikarthik ! Then I think we'll should use a different approach and provide you a way to customize the existing HPA in the chart so it's less opinionated. What do you think?

The PR was updated now, with making changes to existing HPA. Please review!

File: bitnami/thanos/values.yaml [hunk 1]
```diff
@@ -714,13 +714,15 @@ query:
   ## @param query.autoscaling.maxReplicas Maximum number of Thanos Query replicas
   ## @param query.autoscaling.targetCPU Target CPU utilization percentage
   ## @param query.autoscaling.targetMemory Target Memory utilization percentage
+  ## @param query.autoscaling.targetPodMetrics Target Pod metric name and AverageValue
```
Comment (juan131) [inline bitnami/thanos/values.yaml:717]: Given a specific object array structure is expected, we should improve the parameter definition and add some example:
```suggestion
  ## @param query.autoscaling.targetPodMetrics [array] List of custom pod metrics name and averageValue to evaluate for scaling
  ## e.g.
  ## targetPodMetrics:
  ## - name: thanos_query_range_query_duration
  ##   averageValue: 2
```

**Review by juan131 — CHANGES_REQUESTED** [review]
Thanks so much for adapting the PR! Please check my comments.

**Review by juan131 — APPROVED** [review]
LGTM! Thanks for the efforts

**Pull request merged**